tests/serializers/test_any.py

```diff
@@ -177,28 +177,139 @@ def test_any_with_timedelta_serializer():
     ]
 
 
-def test_any_config_timedelta_float():
-    s = SchemaSerializer(core_schema.any_schema(), config={'ser_json_timedelta': 'float'})
-    h2 = timedelta(hours=2)
-    assert s.to_python(h2) == h2
-    assert s.to_python(h2, mode='json') == 7200.0
-    assert s.to_json(h2) == b'7200.0'
-
-    assert s.to_python({h2: 'foo'}) == {h2: 'foo'}
-    assert s.to_python({h2: 'foo'}, mode='json') == {'7200': 'foo'}
-    assert s.to_json({h2: 'foo'}) == b'{"7200":"foo"}'
-
-
-def test_any_config_timedelta_float_faction():
-    s = SchemaSerializer(core_schema.any_schema(), config={'ser_json_timedelta': 'float'})
-    one_half_s = timedelta(seconds=1.5)
-    assert s.to_python(one_half_s) == one_half_s
-    assert s.to_python(one_half_s, mode='json') == 1.5
-    assert s.to_json(one_half_s) == b'1.5'
+@pytest.mark.parametrize(
+    'td,expected_to_python,expected_to_json,expected_to_python_dict,expected_to_json_dict,mode',
+    [
+        (timedelta(hours=2), 7200000.0, b'7200000.0', {'7200000': 'foo'}, b'{"7200000":"foo"}', 'milliseconds_float'),
+        (
+            timedelta(hours=-2),
+            -7200000.0,
+            b'-7200000.0',
+            {'-7200000': 'foo'},
+            b'{"-7200000":"foo"}',
+            'milliseconds_float',
+        ),
+        (timedelta(seconds=1.5), 1500.0, b'1500.0', {'1500': 'foo'}, b'{"1500":"foo"}', 'milliseconds_float'),
+        (timedelta(seconds=-1.5), -1500.0, b'-1500.0', {'-1500': 'foo'}, b'{"-1500":"foo"}', 'milliseconds_float'),
+        (timedelta(microseconds=1), 0.001, b'0.001', {'0.001': 'foo'}, b'{"0.001":"foo"}', 'milliseconds_float'),
+        (
+            timedelta(microseconds=-1),
+            -0.0010000000000287557,
+            b'-0.0010000000000287557',
```

Comment:
This result seems surprising. Have you got any intuition for where the loss of precision is being introduced?

EDIT: I think it's due to the use of floating-point for calculating total seconds. I think we can probably improve the situation by being smarter about having both of `.total_seconds()` and `.total_milliseconds()` methods and being smart about when we do the float division. It seems to me that the second float multiplication (or division) is causing the error.

See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f8e7e433298810b7d40d0a1bc1c7e7c4

Reply:
A couple of these seem to now be solved, however 2 are unsolved, i'll comment on these